lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Random.hs

```diff
@@ -128,10 +150,189 @@ hashSeed :: ScrubbedBytes -> ScrubbedBytes
+     CBOR.encodeWord32 addrIx <>
+     CBOR.encodeBreak
+
+-- | Decode the random HD derivation path section of an address. It will fail to
+-- parse if the address is not in the random scheme.
+-- If the public key is incorrect, the decode result will be 'Nothing'.
+decodeDerivationPath
+    :: Key 'RootK XPub
+    -> CBOR.Decoder s (Maybe (Index 'Hardened 'AccountK, Index 'Soft 'AddressK))
+decodeDerivationPath rootKey = do
+    _ <- CBOR.decodeListLenCanonicalOf 3
+    _ <- CBOR.decodeBytes
+    len <- CBOR.decodeMapLen
+    case len of
+        1 -> do
+            _ <- CBOR.decodeWord8
+            payload <- decodeNestedBytes
+            case decryptDerPath passphrase payload of
+                CryptoFailed _ ->
+                    pure Nothing
+                CryptoPassed plaintext ->
+                    Just <$> deserialiseDerPath plaintext
+        _ -> fail $ "decodeDerivationPath: Incorrect map length: " ++ show len
+  where
+    passphrase = hdPassphrase rootKey
+    deserialiseDerPath :: ByteString -> CBOR.Decoder s (Index 'Hardened 'AccountK, Index 'Soft 'AddressK)
+    deserialiseDerPath bytes =
+        case CBOR.deserialiseFromBytes decodeDerPath (BL.fromStrict bytes) of
+        Right (_, result) -> pure result
+        _ -> fail "decodeDerivationPath: unable to decode path"
+
+
+decodeDerPath :: CBOR.Decoder s (Index 'Hardened 'AccountK, Index 'Soft 'AddressK)
+decodeDerPath = decodeListIndef CBOR.decodeWord32 >>= \case
+    [accIx, addrIx] -> pure (Index accIx, Index addrIx)
+    ps -> fail $ "decodeDerPath: Unexpected derivation path length ("
+        ++ (show $ length ps) ++ ")"
+
+  where
+    -- | Decode an arbitrary long list. CBOR introduce a "break" character to
+    -- mark the end of the list, so we simply decode each item until we encounter
+    -- a break character.
+    --
+    -- Example:
+    --
+    --     myDecoder :: CBOR.Decoder s [MyType]
+    --     myDecoder = decodeListIndef decodeOne
+    --       where
+    --         decodeOne :: CBOR.Decoder s MyType
+    --
+    decodeListIndef :: forall s a. CBOR.Decoder s a -> CBOR.Decoder s [a]
+    decodeListIndef decodeOne = do
+        _ <- CBOR.decodeListLenIndef
+        CBOR.decodeSequenceLenIndef (flip (:)) [] reverse decodeOne
+
+-- | Derive a symmetric key for encrypting and authenticating the address
+-- derivation path.
+hdPassphrase :: Key 'RootK XPub -> Passphrase "address-der"
+hdPassphrase (Key rootXPub) = Passphrase $
+    PBKDF2.generate
+    (PBKDF2.prfHMAC SHA512)
+    -- fixme: check if enough rounds -- if running on a public key, it probably doesn't matter.
+    (PBKDF2.Parameters 500 32)
+    (unXPub rootXPub)
+    ("address-hashing" :: ByteString)
+
+{-------------------------------------------------------------------------------
+                    HD payload encryption and authentication
+-------------------------------------------------------------------------------}
+
+cardanoNonce :: ByteString
+cardanoNonce = "serokellfore"
```

Comment:
that's interesting

Reply:
And that's only the tip of the iceberg my dear :fire: